bundles/org.openhab.persistence.rrd4j/src/main/java/org/openhab/persistence/rrd4j/internal/RRD4jPersistenceService.java

```diff
@@ -330,7 +330,7 @@ public Iterable<HistoricItem> query(FilterCriteria filter) {
                 ts += step;
             }
             return items;
-        } catch (IOException e) {
+        } catch (IllegalArgumentException | IOException e) {
```

Comment:
Wouldn't it be better to not call with invalid parameters? It seems we are simply passing an interval that doesn't make sense here, because `end` could then become smaller than `start`?
```java
long end = filter.getEndDate() == null ? System.currentTimeMillis() / 1000
                : filter.getEndDate().toInstant().getEpochSecond();
```

Reply:
Yes, it would. Do you have a suggestion what we can do? Compare `start` and `end` against each other and switch values if `end` is less than `start`? Or skip querying the database?

Reply:
The equivalent of the current fix would be around line 317 to add:
```java
if (start > end) {
    logger.warn("Could not query rrd4j database for item '{}': start > end", itemName);
    return List.of();
}
```

i.e. avoid calling with invalid parameters. That would be the minimal fix.

Now, the question is which logic is actually desired for missing end date in the filter? For JDBC, missing end date is not interpreted as current date - the condition is simply left out of the query:
https://github.com/openhab/openhab-addons/blob/80094b5e9d293dfb73c81377325da13172d6baaf/bundles/org.openhab.persistence.jdbc/src/main/java/org/openhab/persistence/jdbc/internal/db/JdbcBaseDAO.java#L543-L556

I don't know what would be correct for this service.

Reply:
Me neither. That's the reason why I just added the catch for IAE.

Reply:
Hmm, the intended behavior seems undocumented: https://www.openhab.org/javadoc/latest/org/openhab/core/persistence/filtercriteria

But this also means that it's not documented that end date being null should be interpreted as current date. Since persistence currently only represents historical values (i.e. in the past), a filter with a future start date should not produce any results. So the end result should be the same no matter if we treat null as some distant future date and call, or if we return fast without any results. And I'm not even sure if it makes sense to log a warning.

Do you know the use-case - can you reproduce it, or did any user report this? It would be interesting knowing what lead to this.

Reply:
The exception has been found while testing https://github.com/openhab/openhab-webui/pull/1634. The REST API for querying historic item states provides a `boundary` parameter to return an additional value before and after the given time range. If one tries to fetch data for the current month from RRD4J persistence the IAE occures. Persistence Resource in OHC creates a `FilterCriteria` without an end date to get the value after the given time range.

https://github.com/openhab/openhab-core/blob/8f9dafd1d68a92248cd39414145175952d61cd46/bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/persistence/PersistenceResource.java#L346-L351

I have the slight feeling that the IAE in RRD4J can be forced by calling `PersistenceExtensions.minimumBetween()` or `PersistenceExtensions.maximumBetween()` methods from rules with parameters `begin > end`. OHC does not check that.

Reply:
It seems there is room for improvements in core also (at least documentation) - thanks for the link. I might pick this up when having a bit of time. Reading that code seems to confirm that `end == null` means "no end date" (see difference between these methods):
https://github.com/openhab/openhab-core/blob/dd756b8e2505c6cf9c7555ec7cb73965fd3a5328/bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java#L420-L451

So for this PR, perhaps the solution would be to treat no end date as `Instant.MAX.getEpochSecond()` rather than `System.currentTimeMillis().` Something like (line 283-284):

```java
long end = (filter.getEndDate() == null ? Instant.MAX : filter.getEndDate().toInstant()).getEpochSecond();
```

Reply:
I incorporated the remarks and no longer catch the IAE thrown.

My feeling regarding IAE reproducible via `PersistenceExtension` methods have been confirmed in
https://github.com/openhab/openhab-webui/pull/1634#issuecomment-1397744206.

Reply:
@cweitkamp @jlaur - I was just testing this change and found [this](https://github.com/openhab/openhab-webui/pull/1649#issuecomment-1407621529) - while digging a bit, it seems internally `fetchData` uses something like
```
        int ptsCount = (int) ((fetchEnd - fetchStart) / arcStep + 1);
        long[] timestamps = new long[ptsCount];
```
and since now `fetchEnd` is set to MAX we get OOM. I tried (locally) changing back to `System.currentTimeMillis() / 1000` and OOM is gone. I can create a PR with the fix later today, please let me know what you think, thank you!

Reply:
I was actually afraid something like that could happen because of this very high value of MAX and not knowing the implementation on "the other side", but assumed this would be detected during testing. The logic seems correct IMHO, but probably implementation needs to be adapted. If you could propose a fix, that would be awesome.